components/swim/src/error.rs

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) 2016 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use protobuf;
+
+use std::error;
+use std::fmt;
+use std::io;
+use std::result;
+
+pub type Result<T> = result::Result<T, Error>;
+
+#[derive(Debug)]
+pub enum Error {
+    BadMessage(String),
+    IO(io::Error),
+    ProtobufError(protobuf::ProtobufError),
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let msg = match *self {
+            Error::BadMessage(ref err) => format!("Bad Message: {:?}", err),
+            Error::IO(ref err) => format!("IO Error: {}", err),
+            Error::ProtobufError(ref err) => format!("ProtoBuf Error: {}", err),
+        };
+        write!(f, "{}", msg)
+    }
+}
+
+impl error::Error for Error {
+    fn description(&self) -> &str {
+        match *self {
+            Error::BadMessage(ref _err) => "Bad Protobuf Message; should be Ping/Ack/PingReq",
+            Error::IO(ref err) => err.description(),
+            Error::ProtobufError(ref err) => err.description(),
+        }
+    }
+}
+
+impl From<io::Error> for Error {
+    fn from(err: io::Error) -> Error {
+        Error::IO(err)
```

Comment:
We've got this pattern throughout our codebase but it's the source of our most indestinguishable error messages. An IO Error from the IO module doesn't include any contextual information about where/what went wrong, so I think it's not appropriate to call `Error::from(T)` on these error types.

I'd recommend nipping this in the bud before people keep using it by starting here by not implementing `From<io::Error>` for your error type and generating the error messages with more context instead